### ISSUE TITLE
Task-aware Batching Implementation

### ISSUE THREAD

**Issue by MrZilinXiao** [haon-chen/MoCa#3]
Hi Haonan, 

It's me again. Sorry to raise two issues at the same time, but we are very interested in your work and will definitely cite this amazing work and acknowledge your contribution in our future submissions. 

https://github.com/haon-chen/MoCa/blob/fb3ac280cb69ddb0e769e3903a77e4e6441aa278/src/dataset.py#L139

I notice the current TaskBenchDataset only ensures the length of the entire dataset could be divided by the per-device batch size, but it seems the sampler implementation is missing in order for DataLoader to correctly select batches from the same subset according to the returned `subset_idx`. Do we still use it in the final version of the MoCa model?

**Comment by haon-chen** [comment]
Hi Zilin,

Thanks a lot for the close reading and for planning to cite MoCa—really appreciate it! 🙏

### Why an extra sampler isn’t required

`TaskBatchDataset.initialize_dataset_mapping()` **pads every subset to a multiple of `batch_size` and writes the indices sequentially**:

```
AAA…A | BBB…B | CCC…C
```

The DataLoader simply walks this list in order, so each slice of length `batch_size` stays inside one block ⇒ **all items in a mini-batch originate from the same subset**.

### When would you need a custom sampler?

Only if you set `shuffle=True` or create a sampler that mixes indices inside blocks.

Let me know if anything’s still unclear!

**Comment by lcxrocks** [comment]
Thank you for the detailed discussion above! However, should we consider adding `training_args.shuffle = True` to `cl_train.py`? Without this, the sampling will be completely random across all datasets, leading to a lack of task-aware batching.

**Comment by lcxrocks** [comment]
> Thank you for the detailed discussion above! However, should we consider adding `training_args.shuffle = True` to `cl_train.py`? Without this, the sampling will be completely random across all datasets, leading to a lack of task-aware batching.

I apologize for the misunderstanding. I initially thought that the `training_args.shuffle` argument was a switch for controlling whether random sampling is used in the code, but I did not see any related code. Moreover, I have discovered that this argument does not exist in the `TrainingArguments` class and is likely not implemented by the `Trainer` ([related discussion here])(https://discuss.huggingface.co/t/non-shuffle-training/6986/2). Therefore, could you please further elaborate on how the following behaviour is achieved? 

> Hi Zilin,
> 
> Thanks a lot for the close reading and for planning to cite MoCa—really appreciate it! 🙏
> 
> ### Why an extra sampler isn’t required
> `TaskBatchDataset.initialize_dataset_mapping()` **pads every subset to a multiple of `batch_size` and writes the indices sequentially**:
> 
> ```
> AAA…A | BBB…B | CCC…C
> ```
> 
> The DataLoader simply walks this list in order, so each slice of length `batch_size` stays inside one block ⇒ **all items in a mini-batch originate from the same subset**.
> 
> ### When would you need a custom sampler?
> Only if you set `shuffle=True` or create a sampler that mixes indices inside blocks.
> 
> Let me know if anything’s still unclear!
